zum.1
.TH ZUM 1 "October 26, 2017"
.SH NAME
zum \- free disk space by making holes in files
.SH SYNOPSIS
.B zum
[\fIFILE\fR]...
.SH DESCRIPTION
.B zum
attempts to make the given files sparse.
A sparse file replaces long series of NUL bytes by a count.
This works on most file systems, using \fBcp(1)\fR's \fB\-\-sparse\fR
option.
Note that since \fBzum\fR must make a copy of each file, some free space is
needed for it to work.
.PP
If no arguments are given, \fBzum\fR takes a list of files to process on
\fBstdin\fR, one per line.
.SH EXAMPLE
.IP
find . -type f -print0 | xargs -0 zum
.PP
.SH CAVEATS
If you use LILO as your bootloader, and run \fBzum\fR on your kernel image,
you must rerun lilo, or your system will be unbootable.
.PP
If \fBzum\fR runs out of disk space when processing files, just delete all
files that end with \fB__zum__\fR:
.IP
find DIRECTORY -name '*__zum__' -delete
.PP
After you free some space, it's safe to run it from the beginning again.
.SH SEE ALSO
.BR cp (1),
.BR finddup (1).
.SH AUTHOR
Oleg Kibirev <oleg@gd.cs.CSUFresno.edu>.
.PP
Manual page by Heiko Schlittermann <heiko@lotte.sax.de>.


debian/changelog
finddup (2.0-1) unstable; urgency=low

  * New upstream version. (Closes: #564302, #566998, #449602, #406096).
      - Rewritten finddup either outputs YAML, for ease of
        processing, or shell commands to link duplicate files together.
      - There is no longer a nodup utility/mode.
      - Zum now uses cp to make the file sparse.
      - Package renamed to reflect finddup being the main interest.
      - New utility findcore.

 -- Reuben Thomas <rrt@sc3d.org>  Thu, 26 Oct 2017 10:34:25 +0100

perforate (1.2-5.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Bumped DH level to 10.
  * debian/compat: created. (Closes: #829187)
  * debian/control:
      - Added a Homepage field.
      - Bumped Standards-Version to 3.9.8.
      - Added the ${misc:Depends} variable to Depends field.

 -- Joao Eriberto Mota Filho <eriberto@debian.org>  Wed, 30 Nov 2016 14:22:49 -0200

perforate (1.2-5) unstable; urgency=low

  * Rebuild with unstable libc (Closes: #383590).
  * Added myself to Uploaders.

 -- Hector Garcia <hector@debian.org>  Fri, 18 Aug 2006 22:40:14 +0200

perforate (1.2-4) unstable; urgency=low

  * Fix the the $mode, $uid, $gid information, patch thanks to Philipp
    Matthias Hahn <pmhahn@debian.org> (closes: #356515). 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Thu, 17 Aug 2006 12:43:13 +0200

perforate (1.2-3) unstable; urgency=low

  * Upgraded to DH_COMPAT=5. 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun, 30 Jul 2006 02:04:40 +0200

perforate (1.2-2) unstable; urgency=low

  * Apply patch from Arnaud Fontaine, to deal with FTBFS on hurd-i386 due to
    MAXPATHLEN issue (Closes: #356058).
  * Updated manpage for finddup (closes: #355964).
  * Improved several documentation details (Closes: #375272).
  * Updated Standards version, no changes needed.

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Fri, 28 Jul 2006 09:54:35 +0200

perforate (1.2-1) unstable; urgency=low

  * New upstream (me) release.
  * The "With a little help from my friends" Release.
  * Implement support for large files in zum.c. Patch by Wouter Verhelst.
    (Closes: #255457). Also reimplemnet previously patch by Miriam to fix gcc4
    build warnings. You dude and dudette rock!

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun, 27 Nov 2005 16:39:50 +0100

perforate (1.1-8) unstable; urgency=low

  * Patch provided by Miriam Ruiz, to get rid of gcc4 warnings building zum.c

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sat, 26 Nov 2005 20:09:52 +0100

perforate (1.1-7) unstable; urgency=low

  * The "I should test patches before I upload" Release.
  * Fix nasty syntax error (I introduced myself, 100% unassisted, and quite
    successfully, I must admit) in finddup, with patch from Hector García.
    You know I love you too :P

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sat, 26 Nov 2005 01:07:29 +0100

perforate (1.1-6) unstable; urgency=low

  * The "/me waves at #wave" Release.
  * Take permissions into account, with the optional -i, --ignore-perms
    command line switch. Patch provided by Kari Pahula (Closes: #263782).
    The patch also implements some performance considerations suggested by
    #314548.

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Thu, 10 Nov 2005 12:56:49 +0100

perforate (1.1-5) unstable; urgency=low

  * Improve zum manual page: Add parameters to the SYNOPSIS and ident the
    example. Patch provided by A Costa (Closes: #337745). 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun,  6 Nov 2005 10:13:57 +0100

perforate (1.1-4) unstable; urgency=low

  * Added a warning in zum man page and a README.Debian that explains the
    dangers of zum-ming your /boot directory if using lilo (Closes: #295762). 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Fri,  4 Nov 2005 13:22:25 +0100

perforate (1.1-3) unstable; urgency=low

  * Bumped standars Version to 3.6.2.0, no changes needed.
  * Install a "/usr/bin/nodup" symlink so that invoking "finddup" as "nodup"
    works, and linked also both man pages (Closes: #314264).
  * Add support to define the dirs finddup should look into, patch by my good
    friend Hector Garcia (Closes: #337402).
  * Fix typos in sources (including help and manpage), patch by James R. Van
    Zandt (Closes: #314544).

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Fri,  4 Nov 2005 12:24:37 +0100

perforate (1.1-2) unstable; urgency=low

  * New version of finddup

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Mon,  7 Feb 2005 03:46:37 +0100

perforate (1.1-1) unstable; urgency=low

  * New release.
  * Rewrite of finddup in perl so that it handles espaces in filenames.
    Patch courtesy of Klaus Ethgen <Klaus@Ethgen.de> 
    (Closes: #222030, #263779, #289911, #293790).

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun,  6 Feb 2005 17:26:55 +0100

perforate (1.0-17) unstable; urgency=low

  * Removed Ian Murdock from the Uploaders field. 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun, 19 Dec 2004 01:54:23 +0100

perforate (1.0-16) unstable; urgency=low

  * The compulsive hyperactive housekeeping release.
  * Updated Standards version and DH_COMPAT to 4.
  * Added minor patch to properly handle hardlinks (Closes: #212228).
    Thanks to David Andel <andel@ifi.unizh.ch> and <deblists@osg.saic.com>.
  * Updated changelog to UTF-8. Finally.
  * Added Ian Murdock to the Uploaders field.
  * Fixed zum manpage. Thanks to Heiko Schlittermann <heiko@schlittermann.de>.
    (Closes: #212916).

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Fri, 27 Aug 2004 21:09:01 +0200

perforate (1.0-15) unstable; urgency=low

  * Finddup is now nice to filenames with spaces in them. Many thanks to Diego
    Alvarez <diego.alvarez@entelchile.net> for his patch. (Closes: #152825)

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sat, 27 Jul 2002 06:16:17 +0200

perforate (1.0-14) unstable; urgency=low

  * Corrected README.perforate (Closes: #152813) 

 -- Amaya Rodrigo Sastre <amaya@debian.org>  Sun, 14 Jul 2002 20:00:22 +0200

perforate (1.0-13) unstable; urgency=low


zum.c
/*
 * zum 1.00 - free more disk space by making holes in files.
 *
 * Oleg Kibirev * April 1995 * oleg@gd.cs.CSUFresno.EDU
 * 2005-11-11: Wouter Verhelst <wouter@debian.org>: cleanup
 * 2006-03-10: Arnaud Fontaine <arnau@hurdfr.org>: replace fgets by
 * getline in main function
 * 2017-10-25: Reuben Thomas <rrt@sc3d.org>: assume getline is available,
 * as it's part of POSIX-1.2008.
 *
 * This code is covered by General Public License, version 2 or any later
 * version of your choice. You should recieve file "COPYING" which contains
 * text of the license with any distribution of this program; if you don't 
 * have it, a copy is available from ftp.gnu.ai.mit.edu.
 */

#define _FILE_OFFSET_BITS 64
#define _LARGEFILE_SOURCE
#define _GNU_SOURCE

#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/mman.h>
#include <sys/param.h>
#include <fcntl.h>
#include <errno.h>
#include <utime.h>
#include <unistd.h>
#include <stdlib.h>

static char suffix[] = "__zum__";

static void* my_mmap(void *ptr, int fd, off_t size, off_t *pos)
{
  if(size-(*pos) > (off_t)1<<30) {
    size=1<<30;
  } else {
    size=size-(*pos);
  }
  if(ptr)
    munmap(ptr, 1<<30);
  ptr=mmap(NULL, size, PROT_READ, MAP_PRIVATE, fd, (*pos));
  (*pos)+=size;
  return ptr;
}

static int zero_copy(int fds, int fdd, off_t size)
{
  char *ms;
  char *bp, *p, *ep;
  off_t pos=0;
  int offset;
  
  lseek(fdd, 0L, SEEK_SET);
  if(ftruncate(fdd, 0) < 0) {
    perror("ftruncate");
    return -1;
  }
  
  offset = (size > 1<<30) ? 1<<30 : size;
  if((ms = my_mmap(NULL, fds, size, &pos)) == MAP_FAILED) {
    perror("mmap");
    return -1;
  }

  p = ms; ep = ms + offset;
  
  while(p < ep) {
    for(bp = p; p < ep && *p; p++);
    if(p != bp && write(fdd, bp, p-bp) != p-bp) {
	perror("write");
	munmap(ms, offset);
	return -1;
      }
    for(bp = p; p < ep && !*p; p++);
    if(p != bp)
      lseek(fdd, p-bp, SEEK_CUR);
    if((p == ep) && (size > 1<<30) && (size != pos)) {
      offset = ((size - pos) > 1<<30) ? 1<<30 : (size - pos);
      if((ms = my_mmap(ms, fds, size, &pos)) == MAP_FAILED) {
	perror("mmap");
	return -1;
      } else {
	p = ms; ep = ms + offset;
      }
    }
  }
  munmap(ms, size);
  return ftruncate(fdd, size);
}


static void zero_unmap(char *file)
{
  static int fds = -1, fdd = -1;
  struct stat st, std;

  if(fds > 0) { close(fds); fds = -1; }
  if(fdd > 0) { close(fdd); fdd = -1; }

  if((fds = open(file, O_RDONLY)) < 0 || lstat(file, &st) < 0) {
    perror(file);
    return;
  }

  if(!S_ISREG(st.st_mode)) {
    fprintf(stderr, "%s: Not a regular file.\n", file);
    return;
  }

  fputs(file, stdout);

  char dest[strlen(file) + sizeof(suffix)];
  sprintf(dest, "%s%s", file, suffix);
  if((fdd = open(dest, O_RDWR|O_CREAT|O_EXCL, 0600)) < 0) {
    perror(dest);
    return;
  }

  /* Try to copy the file, and give up if we fail */
  if(zero_copy(fds, fdd, st.st_size) < 0)
    return;

  /* If we can't stat the new file, give up */
  if(fstat(fdd, &std) < 0) {
    perror(dest);
    unlink(dest);
    return;
  }

  /* Check new file is not larger than old one */
  if(std.st_blocks >= st.st_blocks) {
    unlink(dest);
    putchar('\n');
    return;
  }

  printf(" [%uK] ", (unsigned int)((st.st_blocks-std.st_blocks)*st.st_blksize/1024));

  if(st.st_nlink == 1) {
    unlink(file);
    if(rename(dest, file) < 0) {
      perror("rename");
      return;
    }
    fputs(" [1 link]\n", stdout);
  } else {
    close(fds);
    if((fds = open(file, O_RDWR)) < 0) {
      perror(file);
      unlink(dest);
      return;
    }
    if(zero_copy(fdd, fds, st.st_size) < 0)
      return;

    unlink(dest);
    fputs(" [many links]\n", stdout);
  }

  /* 
   * Don't check chown errors, since after all this might not be root.
   */
  chown(file, st.st_uid, st.st_gid);
  if(chmod(file, st.st_mode) < 0)
    perror("Can't reset mode");
  
  {
    struct utimbuf b;
    b.actime  = st.st_atime;
    b.modtime = st.st_mtime;
    utime(file, &b);
  }
}

int main(int argc, char **argv)
{
  char *p;

  setbuf(stdout, NULL); /* Make stdout unbuffered */

  if(argc > 1)
    while((p = *(++argv)))
      zero_unmap(p);
  else {
    char *buf = NULL;
    size_t len = 0;
    while(getline(&buf, &len, stdin) != -1)
      zero_unmap(buf);
    
    if (buf)
      free(buf);
  }
  return 0;
}
